framework/src/meshgenerators/Poly2TriMeshGenerator.C

```diff
@@ -0,0 +1,245 @@
+    _refine_holes(getParam<std::vector<bool>>("refine_holes")),
+    _desired_area(getParam<Real>("desired_area")),
+    _desired_area_func(getParam<std::string>("desired_area_func")),
+    _algorithm(parameters.get<MooseEnum>("algorithm")),
+    _verbose_stitching(parameters.get<bool>("verbose_stitching"))
+{
+  if (!_stitch_holes.empty() && _stitch_holes.size() != _hole_ptrs.size())
+    paramError("stitch_holes", "Need one stitch_holes entry per hole, if specified.");
+
+  for (auto hole_i : index_range(_stitch_holes))
+    if (hole_i < _refine_holes.size() && _refine_holes[hole_i])
+      paramError("refine_holes", "Disable auto refine of any hole boundary to be stitched.");
+}
+
+std::unique_ptr<MeshBase>
+Poly2TriMeshGenerator::generate()
+{
+  // We put the boundary mesh in a local pointer
+  std::unique_ptr<UnstructuredMesh> mesh = dynamic_pointer_cast<UnstructuredMesh>(_bdy_ptr);
+
+  Poly2TriTriangulator poly2tri(*mesh);
+  poly2tri.triangulation_type() = TriangulatorInterface::PSLG;
+
+  poly2tri.set_interpolate_boundary_points(_add_nodes_per_boundary_segment);
+  poly2tri.set_refine_boundary_allowed(_refine_bdy);
+
+  poly2tri.desired_area() = _desired_area;
+  poly2tri.minimum_angle() = 0; // Not yet supported
+  poly2tri.smooth_after_generating() = _smooth_tri;
+
+  if (_desired_area_func != "")
+  {
+    // poly2tri will clone this so it's fine going out of scope
+    ParsedFunction<Real> area_func{_desired_area_func};
+    poly2tri.set_desired_area_function(&area_func);
+  }
+
+  std::vector<TriangulatorInterface::MeshedHole> meshed_holes;
+  std::vector<TriangulatorInterface::Hole *> triangulator_hole_ptrs;
+
+  // Make sure pointers here aren't invalidated by a resize
+  meshed_holes.reserve(_hole_ptrs.size());
+  for (auto hole_i : index_range(_hole_ptrs))
+  {
+    meshed_holes.emplace_back(*_hole_ptrs[hole_i]->get());
+    if (hole_i < _refine_holes.size())
+      meshed_holes.back().set_refine_boundary_allowed(_refine_holes[hole_i]);
+
+    triangulator_hole_ptrs.push_back(&meshed_holes.back());
+  }
+
+  if (!triangulator_hole_ptrs.empty())
+    poly2tri.attach_hole_list(&triangulator_hole_ptrs);
+
+  poly2tri.triangulate();
+
+  // I do not trust Laplacian mesh smoothing not to invert elements
+  // near reentrant corners.  Eventually we'll add better smoothing
+  // options, but even those might have failure cases.  Better to
+  // always do an extra loop here than to ever let users try to run on
+  // a degenerate mesh.
+  if (_smooth_tri)
+    for (auto elem : mesh->element_ptr_range())
+    {
+      mooseAssert(elem->type() == TRI3, "Unexpected non-Tri3 found in triangulation");
+      auto cross_prod = (elem->point(1) - elem->point(0)).cross(elem->point(2) - elem->point(0));
+
+      if (cross_prod(2) <= 0)
+        mooseError("Inverted element found in triangulation.\n"
+                   "Laplacian smoothing can create these at reentrant corners; disable it?");
+    }
```

Comment:
should you add this to the SmoothMeshGenerator we already have then

Reply:
I didn't know that existed, but sure.  Obviously not in this PR, though.